src/Analyzer/MatcherNode.h

```diff
@@ -0,0 +1,179 @@
+  * Example: SELECT compound_column.* FROM test_table.
+  *
+  * If identifier is not resolved as expression try to resolve it as table.
+  * If identifier is resolved as table then apply matcher to it.
+  * Example: SELECT test_table.* FROM test_table.
+  * Example: SELECT a.* FROM test_table AS a.
+  *
+  * Additionally each matcher can contain transformers, check ColumnTransformers.h.
+  * In query tree matchers column transformers are represended as ListNode.
+  */
+enum class MatcherNodeType
+{
+    ASTERISK,
+    COLUMNS_REGEXP,
+    COLUMNS_LIST
+};
+
+const char * toString(MatcherNodeType matcher_node_type);
+
+class MatcherNode;
+using MatcherNodePtr = std::shared_ptr<MatcherNode>;
+
+class MatcherNode final : public IQueryTreeNode
+{
+public:
+    /// Variant unqualified asterisk
+    explicit MatcherNode(ColumnTransformersNodes column_transformers_ = {});
+
+    /// Variant qualified asterisk
+    explicit MatcherNode(Identifier qualified_identifier_, ColumnTransformersNodes column_transformers_ = {});
+
+    /// Variant unqualified COLUMNS('regexp')
+    explicit MatcherNode(std::shared_ptr<re2::RE2> columns_matcher_, ColumnTransformersNodes column_transformers_ = {});
+
+    /// Variant qualified COLUMNS('regexp')
+    explicit MatcherNode(Identifier qualified_identifier_, std::shared_ptr<re2::RE2> columns_matcher_, ColumnTransformersNodes column_transformers_ = {});
+
+    /// Variant unqualified COLUMNS(column_name_1, ...)
+    explicit MatcherNode(Identifiers columns_identifiers_, ColumnTransformersNodes column_transformers_ = {});
+
+    /// Variant qualified COLUMNS(column_name_1, ...)
+    explicit MatcherNode(Identifier qualified_identifier_, Identifiers columns_identifiers_, ColumnTransformersNodes column_transformers_ = {});
+
+    /// Get matcher type
+    MatcherNodeType getMatcherType() const
+    {
+        return matcher_type;
+    }
+
+    /// Is this matcher represented by asterisk
+    bool isAsteriskMatcher() const
+    {
+        return matcher_type == MatcherNodeType::ASTERISK;
+    }
+
+    /// Is this matcher represented by COLUMNS
+    bool isColumnsMatcher() const
+    {
+        return matcher_type == MatcherNodeType::COLUMNS_REGEXP || matcher_type == MatcherNodeType::COLUMNS_LIST;
+    }
+
+    /// Returns true if matcher qualified with identifier, false otherwise
+    bool isQualified() const
+    {
+        return !qualified_identifier.empty();
+    }
+
+    /// Returns true if matcher is not qualified with identifier, false otherwise
+    bool isUnqualified() const
+    {
+        return qualified_identifier.empty();
+    }
+
+    /// Get qualified identifier
+    const Identifier & getQualifiedIdentifier() const
+    {
+        return qualified_identifier;
+    }
+
+    /// Get columns matcher. Valid only if this matcher has type COLUMNS_REGEXP.
+    const std::shared_ptr<re2::RE2> & getColumnsMatcher() const
+    {
+        return columns_matcher;
```

Comment:
I will make matcher node refactoring in separate pull request. I expect that because we do not have check here, it is possible that some logic check matcher type using this method.